### PR TITLE
Add support for X.509 auth for HTTP and MQTT over Websockets

### PR DESCRIPTION
**Pull request by mrohera**

### CODE REVIEW

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Http/middleware/TlsConnectionFeatureExtended.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+//----------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  All rights reserved.
+//----------------------------------------------------------------
+
+namespace Microsoft.Azure.Devices.Edge.Hub.Http.Middleware
+{
+    using System.Collections.Generic;
+    using System.Security.Cryptography.X509Certificates;
+
+    public class TlsConnectionFeatureExtended : ITlsConnectionFeatureExtended
```
Comment (varunpuranik) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Http/middleware/TlsConnectionFeatureExtended.cs]: I would like to understand this a bit more.. where is the chain added to the class? If this is expected to be just a holder class, maybe the interface is not needed?

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Http/middleware/WebSocketHandlingMiddleware.cs [hunk 2]
```diff
@@ -66,7 +68,18 @@ async Task ProcessRequestAsync(HttpContext context, IWebSocketListener listener,
                 localEndPoint = Option.Some<EndPoint>(new IPEndPoint(context.Connection.LocalIpAddress, context.Connection.LocalPort));
             }
             var remoteEndPoint = new IPEndPoint(context.Connection.RemoteIpAddress, context.Connection.RemotePort);
-            await listener.ProcessWebSocketRequestAsync(webSocket, localEndPoint, remoteEndPoint, correlationId);
+            var cert = await context.Connection.GetClientCertificateAsync();
+            Option<X509Certificate2> clientCertificate = (cert == null) ? Option.None<X509Certificate2>() :
```
Comment (varunpuranik) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Http/middleware/WebSocketHandlingMiddleware.cs]: Option.Maybe

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Http/middleware/ConnectionInfoExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace Microsoft.Azure.Devices.Edge.Hub.Http.Middleware
+{
+    using Microsoft.AspNetCore.Http;
+    using System.Collections.Generic;
+    using System.Security.Cryptography.X509Certificates;
+
+    public static class ConnectionInfoExtensions
+    {
+        public static IList<X509Certificate2> GetClientCertificateChain(this ConnectionInfo connectionInfo, HttpContext context)
```
[the remark this answers is not in the document]
Comment (varunpuranik) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Http/middleware/ConnectionInfoExtensions.cs]: I see what you are doing here.. trying to keep consistency with GetClientCertificate? But it seems a bit odd, I would just pass in the HttpContext.. should be fine imo.

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Mqtt/MqttWebSocketListener.cs [hunk 4]
```diff
@@ -66,6 +74,13 @@ public async Task ProcessWebSocketRequestAsync(WebSocket webSocket, Option<EndPo
                     remoteEndPoint
                 );
 
+                clientCert.Map(cert =>
+                {
+                    return identityProvider.RemoteCertificateValidationCallback(cert,
+                                                                                clientCertChain.GetOrElse(() =>
+                                                                                    throw new ArgumentException("Certificate chain was found to be null")));
```
Comment (varunpuranik) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Mqtt/MqttWebSocketListener.cs]: Expect

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Service/Hosting.cs [hunk 5]
```diff
@@ -30,7 +31,11 @@ public class Hosting
                 {
                     options.Listen(!Socket.OSSupportsIPv6 ? IPAddress.Any : IPAddress.IPv6Any, port, listenOptions =>
                     {
-                        listenOptions.UseHttps(serverCertificate);
+                        listenOptions.UseHttps(serverCertificate, (connectionAdapterOptions) =>
+                        {
+                            connectionAdapterOptions.ClientCertificateMode = ClientCertificateMode.AllowCertificate;
+                            connectionAdapterOptions.ClientCertificateValidation = (clientCert, chain, policyErrors) => true;
+                        });
```
Comment (varunpuranik) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Service/Hosting.cs]: So, if we do this, will the cert actually be "verified" (whether it is correct, matches, the chain, etc)?

Comment (mrohera) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Service/Hosting.cs]: Verification will happen in authenticate async in all cases. Essentially we have taken validation out of the TLS (or transport layer) and brought it up to the application layer.

Comment (mrohera) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Service/Hosting.cs]: I can add a comment to mention this.

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Http/middleware/TlsConnectionFeatureExtended.cs (hunk 1 above)
Comment (varunpuranik) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Http/middleware/TlsConnectionFeatureExtended.cs]: Looks like this has no ref..

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Http/middleware/ConnectionInfoExtensions.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (mrohera) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Http/middleware/ConnectionInfoExtensions.cs]: Thank this was a nice suggestion.

**Comment by mrohera** [comment]
Will add some tests for HttpsExtensionConnectionAdapter

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Http/Adapters/HttpsExtensionConnectionAdapter.cs [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+    using Microsoft.AspNetCore.Http.Features;
+    using Microsoft.AspNetCore.Server.Kestrel.Core.Adapter.Internal;
+    using Microsoft.AspNetCore.Server.Kestrel.Https;
+    using Microsoft.Azure.Devices.Edge.Hub.Http.Middleware;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Logging;
+
+    // https://github.com/aspnet/HttpAbstractions/issues/808
+    public class HttpsExtensionConnectionAdapter : IConnectionAdapter
+    {
```
Comment (varunpuranik) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Http/Adapters/HttpsExtensionConnectionAdapter.cs]: Looks like you have modified this file slightly.. in that case we should align it with EdgeHub semantics. Alternatively, I am okay this file being completely unmodified..

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Http/Adapters/HttpsExtensionConnectionAdapter.cs [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+        // See http://oid-info.com/get/1.3.6.1.5.5.7.3.1
+        // Indicates that a certificate can be used as a SSL server certificate
+        private const string ServerAuthenticationOid = "1.3.6.1.5.5.7.3.1";
+        const string AuthenticationSucceeded = "AuthenticationSucceeded";
+        internal const string DisableHandshakeTimeoutSwitch = "Switch.Microsoft.AspNetCore.Server.Kestrel.Https.DisableHandshakeTimeout";
+        private static readonly TimeSpan HandshakeTimeout = TimeSpan.FromSeconds(10);
```
Comment (varunpuranik) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Http/Adapters/HttpsExtensionConnectionAdapter.cs]: remove private everywhere

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Http/middleware/ITlsConnectionFeatureExtended.cs [hunk 8]
```diff
@@ -0,0 +1,12 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+namespace Microsoft.Azure.Devices.Edge.Hub.Http.Middleware
+{
+    using System.Collections.Generic;
+    using System.Security.Cryptography.X509Certificates;
+
+    public interface ITlsConnectionFeatureExtended
+    {
```
[the remark this answers is not in the document]
Comment (mrohera) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Http/middleware/ITlsConnectionFeatureExtended.cs]: This new interface is consistent with how connection features are extended. So I feel there is merit in keeping things consistent.

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Http/middleware/HttpContextExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,17 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+namespace Microsoft.Azure.Devices.Edge.Hub.Http.Middleware
```
[the remark this answers is not in the document]
Comment (mrohera) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Http/middleware/HttpContextExtensions.cs]: Sure I thought this was consumed in middleware and so was a logical home. Where do suggest we add this?

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Http/Adapters/HttpsExtensionConnectionAdapter.cs (hunk 6 above)
Comment (mrohera) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Http/Adapters/HttpsExtensionConnectionAdapter.cs]: This file has not been modified apart from log messages and certificate and a few Edge utilities and Edge hub conventions.

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Http/middleware/HttpContextExtensions.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (mrohera) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Http/middleware/HttpContextExtensions.cs]: Fixed.

**Pull request merged**